### ISSUE TITLE
Cannot attach a debugger to Firebase Python Functions

### ISSUE THREAD

**Issue by johnnyoshika** [firebase/firebase-tools#6838]
### [REQUIRED] Environment info

**firebase-tools:** 12.9.1

**Platform:** Windows / macOS

### [REQUIRED] Test case

Run emulator with `--inspect-functions` flag:

```
firebase emulators:start --project demo-project --inspect-functions
```

### [REQUIRED] Steps to reproduce

Run emulator with `--inspect-functions` flag:

```
firebase emulators:start --project demo-project --inspect-functions
```

This message appears in the console:

```
>  Debugger attached.
!  --inspect-functions not supported for Python functions. Ignored.
```

### [REQUIRED] Expected behavior

Debugger would attach to Python functions

### [REQUIRED] Actual behavior

Receive a message in the Console that debugger is not supported

**Comment by aalej** [comment]
Hey @johnnyoshika, thanks for reaching out. Currently, `--inspect-functions` is only supported for Node.js runtimes. I’m going to label this as a feature request so we can better track this.

For those who also find this feature valuable, please leave a thumbs up on the original post. This helps our engineers prioritize future development efforts.

**Comment by pamafe1976** [comment]
As a workaround, I tried to debug my python functions starting the emulators and then executing the starting the function manually. However for this to work I need to provide credentials, and then the function targets the production services instead of the emulator.
Anyone knows a way of doing initialize_app forcing it to point to the firestore emulator?

**Comment by johnnyoshika** [comment]
> As a workaround, I tried to debug my python functions starting the emulators and then executing the starting the function manually. However for this to work I need to provide credentials, and then the function targets the production services instead of the emulator. Anyone knows a way of doing initialize_app forcing it to point to the firestore emulator?

Have you tried setting these env variables?

```
process.env.FIREBASE_AUTH_EMULATOR_HOST = '127.0.0.1:9099';
process.env.FIRESTORE_EMULATOR_HOST = '127.0.0.1:8080';
```

**Comment by johnnyoshika** [comment]
@pamafe1976 Thank you! I'll give that a try the next time I need to debug these functions.

**Comment by Chezlui** [comment]
@pamafe1976 so the main.py is like a custom one that you have to test these functions, or it is the typical where you declare all the endpoints (@https_fn.on_request ...)?

**Comment by pamafe1976** [comment]
Its actually the same main.py

At the beginning I have a global variable

`RUN_LOCAL = False`

and the body of the main is something like this:

```
def InitLocalDebug() -> None:
    DIRNAME = os.path.dirname(__file__)
    BASEDIR = os.path.abspath(os.path.join(DIRNAME, ".."))
    GOOGLE_APPLICATION_CREDENTIALS = os.path.join(
        BASEDIR,
        "credentials",
        "credential_file.json",
    )
    os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = GOOGLE_APPLICATION_CREDENTIALS
    os.environ["FIREBASE_AUTH_EMULATOR_HOST"] = "192.168.1.93:9099"
    os.environ["FIRESTORE_EMULATOR_HOST"] = "192.168.1.93:8080"
    load_dotenv(os.path.join(DIRNAME, ".env.local"))

if RUN_LOCAL:
    InitLocalDebug()

initialize_app()

if RUN_LOCAL:
    requestLocations("")
```

So when I set RUN_LOCAL=True, I can debug locally the function that is called (in this case requestLocations)
And if I set RUN_LOCAL=False, I can either deploy or run with local emulator

**Comment by Chezlui** [comment]
@aalej to be honest is pretty annoying. 

I don't understand why there are so few resources related to Python environments compared to JS when both platforms are supposed to be supported in Firebase Functions.

Something fails silently and you spend hours adding traces everywhere when debugging you would find it in a matter of minutes.

Are we just 5 people using Python here?

**Comment by pamafe1976** [comment]
I guess it may because Python support started with 2nd gen

**Comment by Chezlui** [comment]
@pamafe1976 I forgot, thank you very much for the tip.
Right now it doesn't work for me, because most of my problems are in the middle of sequences of requests and responses of different functions, so using just one doesn't help me a lot.

But it's really helpful though and I will use it for sure.

**Comment by fatihsever** [comment]
In the AI ​​era, we use Python functions more. It is really challenging not being able to debug these Python functions that we use a lot while developing the AI/ML features of our applications. Could you please prioritize this issue? Thanks.

**Comment by Chezlui** [comment]
@aalej I feel any developer that uses Python with Firebase needs this, don't you? I mean, I understand using thumbs up can be a good way to prioritize, but maybe most of developers don't find this issue.

Just curiosity, how do you sort it out this if you don't have the chance to debug? Pro tips may help others.

**Comment by vikx01** [comment]
At the very least your docs should reflect the fact that currently there is no way to attach a debugger to firebase functions written in python.

**Comment by Chezlui** [comment]
That's true, @vikx01 ,  I spent hours setting up the way to debug and when I got it ready I found out this amazing suprise:

![image](https://github.com/firebase/firebase-tools/assets/4120243/c0ca9150-7e82-474f-b305-3b3738f7d801)

**Comment by fatihsever** [comment]
I totally agree, @vikx01
Anyone who has come to this GitHub issue has probably experienced the following :)

After starting the emulator with --inspect-functions, you began debugging in VSCode as usual. You liked it because the debugging started. Then you realized that it never hits the breakpoint. You spent hours searching for the answer on StackOverflow, ChatGPT, Gemini, Llama, Mistral, and various other LLMs, but couldn't find the correct answer. Finally, after seeing the amazing surprise that @Chezlui mentioned, you gave up and started filling the code with print()s to find the error in the ancient way.

I think `--inspect-functions only supported for Node.js` message deserves more attention than the following warning that's printed hundreds of times in the emulator console:
<img width="844" alt="zsh_—_functions" src="https://github.com/firebase/firebase-tools/assets/729293/e5a6d7f6-23f9-4d31-997f-ccfecb6967b0">

**Comment by ravpiotrowski** [comment]
hello, any news? Is it possible to use debugger in the Cloud Function for the Python environment in 2025? :D I mean not hacking, of course...

**Comment by Christiaan-Mathu** [comment]
Would also like to know please?

**Comment by fatihsever** [comment]
That's why we started a migration project from Firebase Functions to Cloud Run Functions.
It will be useful not only for debugging problems but also for the problems mentioned in this article:

https://surferjeff.medium.com/5-reasons-why-we-moved-to-google-cloud-run-from-firebase-functions-f6aec99a03f9

Thanks to @surferjeff for this useful article.

**Comment by sitanshu10** [comment]
Why is there still no update on this?? We need a debugger ASAP

**Comment by kornha** [comment]
Can we escalate this? I don't see why this is so difficult to implement, given the above code, and this causes huge problems for developers
cc @taeold @aalej thanks!

**Comment by tristan-north** [comment]
Disappointing to get to the end of this thread without any great solutions - anyone have news on python debugging?

**Comment by Bradley-McCallion** [comment]
@taeold Are you accepting contributions for this issue, if nobody is able to tackle this I'd like to give it a go?

**Comment by taeold** [comment]
@Bradley-McCallion yes open to contribution!

Last time I looked, I ended up researching https://github.com/microsoft/debugpy/, but ultimately became distracted w/ something else. Would love to hear what you have in mind.

**Comment by Bradley-McCallion** [comment]
@taeold I just thought about using the NodeJS implementation as a template.

When `--inspect-functions` is used, the CLI sets up the Node.js function process with the Node inspector (https://github.com/firebase/firebase-tools/blob/ece5bcd48e32e22bc3680c172b9f81f11c3238bd/src/emulator/functionsEmulator.ts#L1590-L1598)

On the python side, when `debugMode` is enabled we should spawn the functions framework process with a debugger attached (like you said proabably `debugpy`), `python -m debugpy --listen <port> -m functions_framework ...`. We would probably also just need to ensure that only a single Python function worker handles requests sequentially (similar to what is done on the node side).

Specifically we would probably just look to modify this function (https://github.com/firebase/firebase-tools/blob/ece5bcd48e32e22bc3680c172b9f81f11c3238bd/src/emulator/functionsEmulator.ts#L1639) to something like:

```py
...
const args = [];
if (this.debugMode) {
    let debugPort = ...; 
    args.push("python", "-m", "debugpy", "--listen", `127.0.0.1:${debugPort}`, "-m", "functions_framework");
} else {
    args.push("functions-framework");
}
const childProcess = runWithVirtualEnv(args, backend.functionsDir, {...envs, PORT: port, HOST: "127.0.0.1", ...});
```

In practice we'd obviously need to harden this approach with some error handling and the likes but this was the high level overview I had in mind

**Comment by kornha** [comment]
makes sense to me! @Bradley-McCallion

**Comment by Bradley-McCallion** [comment]
@taeold @kornha I've implemented a workaround for this right now in our setup using the @init decorator to start the debugger. This is good enough for us right now without having the proper support. Apologies I won't have the time to contribute towards fixing this right now

**Comment by taeold** [comment]
@Bradley-McCallion thanks for taking the time to investigate. If you can share your workaround with us here, I'd really appreciate it!

**Comment by Bradley-McCallion** [comment]
@taeold Of course! So in each of my firebase function sources I add the following initialisation callback:

```py
import os
from firebase_functions.params import BoolParam
from firebase_functions.core import init

ENABLE_DEBUGGER = BoolParam("ENABLE_DEBUGGER", default=False)

@init
def initialize() -> None:
    if os.environ.get("DEBUGGER_RUNNING", "") != "true":
        import debugpy
        debugpy.listen(("127.0.0.1", 5678))
        os.environ["DEBUGGER_RUNNING"] = "true"
        print("⚡ debugpy listening on 127.0.0.1:", "5678", flush=True)
        debugpy.wait_for_client()
```

If ENABLE_DEBUGGER is set to true (which I can use a `.env.local` to set and commit to code), then on the first function call this initialisation decorator will be hit, start debugpy listening on port 5678 and then wait for a client connection.

At the same time in my `.vscode/launch.json` I have the following debug launching configuration:
```
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Attach to Firebase Function",
            "type": "debugpy",
            "request": "attach",
            "connect": {
                "host": "127.0.0.1",
                "port": 5678
            }
        }

    ],
}
```

Therefore after running either a first dummy call or starting my integration tests I just need to make sure I see that the debugger has started from the log, then I can launch my debug client and the debugger will attach and the code will continue.

Its not ideal in two ways:
1. I need to pollute my main code with the debug launching setup
2. I need to explicitly launch my debug configuration after the initialisation callback (which is not run on deploying the functions only on function start)

However, it's easy enough and does the job for now.
